File: OTHER_END/mobile/tuan-app/src/utils/robotSimulator.ts
import { robotState, MS_TO_KMH, updateRobotLocation, updateRobotStatus, updateRobotTelemetry } from './robotStore'
import type { RobotState } from './robotStore'

/**
 * 机器人 GPS 模拟器（对齐 API 协议）
 * - 自动模式：沿巡检路线 polyline 匀速移动
 * - 手动模式：响应遥控页面摇杆，8方向自由移动
 * - 遥测：生成 enc1/enc2/steer 编码器数据
 * - 定位：生成 loc 格式 GPS 数据（速度 km/h）
 */

const M_PER_DEG_LAT = 111000
const TICK_MS = 1000
const M_TO_KM = 1000
const DEFAULT_AUTO_SPEED_MS = 1.5
const ENCODER_PULSE_RATE = 65
const STEER_CENTER = 150
const BATTERY_TICK_INTERVAL = 60
const SIGNAL_TICK_INTERVAL = 30
const BATTERY_MIN = 5

/** 8方向航向角映射 */
const HEADING_ANGLES: Record<string, number> = {
  forward: 0, backward: 180, left: 270, right: 90,
  forward_left: 315, forward_right: 45,
  backward_left: 225, backward_right: 135,
}

/** 随机刷新信号强度 */
function randomizeSignal(): void {
  const signals: Array<RobotState['signal']> = ['强', '强', '强', '中', '弱']
  updateRobotStatus({ signal: signals[Math.floor(Math.random() * signals.length)] })
}

let timer: ReturnType<typeof setInterval> | null = null
let pathIndex = 0
let forward = true
let progress = 0
let elapsed = 0
let pulseCount = 0 // 模拟编码器脉冲累计

// ---- 工具 ----

function distMeters(a: { latitude: number; longitude: number }, b: { latitude: number; longitude: number }): number {
  const dLat = (b.latitude - a.latitude) * M_PER_DEG_LAT
  const dLng = (b.longitude - a.longitude) * M_PER_DEG_LAT * Math.cos(((a.latitude + b.latitude) / 2) * (Math.PI / 180))
  return Math.sqrt(dLat * dLat + dLng * dLng)
}

function interpolate(a: { latitude: number; longitude: number }, b: { latitude: number; longitude: number }, t: number) {
  return { latitude: a.latitude + (b.latitude - a.latitude) * t, longitude: a.longitude + (b.longitude - a.longitude) * t }
}

// ---- 主 tick ----

function tick(): void {
  elapsed++
  if (robotState.controlDirection) {
    tickManual()
  } else {
    tickAuto()
  }
  tickTelemetry()
  tickBattery()
}

/** 手动遥控：8方向移动 */
function tickManual(): void {
  const dir = robotState.controlDirection
  const head = HEADING_ANGLES[dir] ?? 0
  const rad = (head * Math.PI) / 180
  const speedMs = robotState.controlSpeed // m/s
  const dist = speedMs
  const cosLat = Math.cos((robotState.latitude * Math.PI) / 180)

  const dLat = (dist * Math.cos(rad)) / M_PER_DEG_LAT
  const dLng = (dist * Math.sin(rad)) / (M_PER_DEG_LAT * cosLat)

  // 更新为 WGS-84 坐标（模拟器直接生成，无需转换）
  updateRobotLocation(
    robotState.latitude + dLat,
    robotState.longitude + dLng,
    head,
    +(speedMs * MS_TO_KMH).toFixed(1),
  )
  updateRobotStatus({ mileage: +(robotState.mileage + dist / M_TO_KM).toFixed(2) })
  updateRobotStatus({ fix: 3, satellites: 12, locSrc: 'gps' })

  // 编码器模拟
  pulseCount += Math.round(speedMs * 100)
  updateRobotTelemetry(pulseCount, pulseCount + Math.round(Math.random() * 5), STEER_CENTER)
}

/** 自动巡航：沿 polyline 移动 */
function tickAuto(): void {
  const path = robotState.path
  if (path.length < 2) return

  const nextIndex = forward ? pathIndex + 1 : pathIndex - 1
  if (nextIndex < 0 || nextIndex >= path.length) {
    forward = !forward
    pathIndex = forward ? 0 : path.length - 1
    return
  }

  const from = path[pathIndex]
  const to = path[nextIndex]
  const segDist = distMeters(from, to)
  const speedMs = robotState.controlDirection ? robotState.controlSpeed : DEFAULT_AUTO_SPEED_MS
  const stepRatio = segDist > 0 ? speedMs / segDist : 0
  progress += stepRatio

  if (progress >= 1) {
    progress = 0
    pathIndex = nextIndex
    robotState.nextPoint = `K12+${300 + (forward ? pathIndex : path.length - 1 - pathIndex) * 150}`
  }

  const pos = interpolate(from, to, Math.min(progress, 1))
  const rawHeading = (Math.atan2((to.longitude - from.longitude) * (forward ? 1 : -1), (to.latitude - from.latitude) * (forward ? 1 : -1)) * 180) / Math.PI + 90
  const heading = ((rawHeading % 360) + 360) % 360

  updateRobotLocation(pos.latitude, pos.longitude, heading, +(speedMs * MS_TO_KMH).toFixed(1))
  updateRobotStatus({ mileage: +(robotState.mileage + speedMs / M_TO_KM).toFixed(2) })

  if (elapsed % SIGNAL_TICK_INTERVAL === 0) {
    randomizeSignal()
  }
}

/** 编码器遥测 */
function tickTelemetry(): void {
  pulseCount += Math.round(robotState.controlSpeed * ENCODER_PULSE_RATE)
  const enc1 = pulseCount
  const isTurningLeft = robotState.controlDirection.includes('left')
  const isTurningRight = robotState.controlDirection.includes('right')
  const enc2 = pulseCount + (isTurningLeft ? -20 : isTurningRight ? 20 : Math.round(Math.random() * 5))
  const steer = isTurningLeft ? 120 : isTurningRight ? 180 : STEER_CENTER
  updateRobotTelemetry(enc1, enc2, steer)
}

function tickBattery(): void {
  if (elapsed % BATTERY_TICK_INTERVAL === 0 && robotState.battery > BATTERY_MIN) {
    updateRobotStatus({ battery: robotState.battery - 1 })
  }
  if (elapsed % SIGNAL_TICK_INTERVAL === 0) {
    randomizeSignal()
  }
}

// ---- 公开 API ----

export function startSimulation(): void {
  if (timer) return
  robotState.connected = true
  robotState.mode = 'simulation'
  robotState.status = 'running'
  robotState.controlSpeed = DEFAULT_AUTO_SPEED_MS
  pathIndex = 0; progress = 0; forward = true; elapsed = 0
  tick()
  timer = setInterval(tick, TICK_MS)
}

export function stopSimulation(): void {
  if (timer) { clearInterval(timer); timer = null }
  robotState.connected = false
  robotState.controlSpeed = 0
  robotState.status = 'online'
  updateRobotTelemetry(0, 0, 150)
}

/** 遥控方向（8方向 + stop） */
export function simSetDirection(dir: string): void {
  if (dir === 'stop' || dir === '') {
    robotState.controlDirection = ''
    robotState.controlSpeed = 0
    robotState.speedKmh = 0
    updateRobotTelemetry(pulseCount, pulseCount, 150)
  } else {
    robotState.controlDirection = dir
    robotState.controlSpeed = 0.8 + Math.random() * 0.6
    robotState.speedKmh = +(robotState.controlSpeed * 3.6).toFixed(1)
  }
}

File: OTHER_END/android_remote/app/src/main/java/com/suoxingtuan/inspector/data/simulator/RobotSimulator.kt
package com.suoxingtuan.inspector.data.simulator

import com.suoxingtuan.inspector.data.model.LatLng
import com.suoxingtuan.inspector.data.repository.RobotRepository
import kotlinx.coroutines.*
import kotlin.math.*

/**
 * Robot GPS simulator — mirrors robotSimulator.ts.
 * Auto mode: follows polyline path. Manual mode: 8-direction free movement.
 */
object RobotSimulator {
    private const val M_PER_DEG_LAT = 111000.0
    private const val TICK_MS = 1000L
    private const val DEFAULT_AUTO_SPEED_MS = 1.5
    private const val ENCODER_PULSE_RATE = 65
    private const val STEER_CENTER = 150
    private const val BATTERY_MIN = 5

    private val HEADING_ANGLES = mapOf(
        "forward" to 0, "backward" to 180, "left" to 270, "right" to 90,
        "forward_left" to 315, "forward_right" to 45,
        "backward_left" to 225, "backward_right" to 135
    )

    private var timer: Job? = null
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private var pathIndex = 0
    private var forward = true
    private var progress = 0.0
    private var elapsed = 0
    private var pulseCount = 0

    private fun distMeters(a: LatLng, b: LatLng): Double {
        val dLat = (b.latitude - a.latitude) * M_PER_DEG_LAT
        val dLng = (b.longitude - a.longitude) * M_PER_DEG_LAT * cos((a.latitude + b.latitude) / 2 * (PI / 180))
        return sqrt(dLat * dLat + dLng * dLng)
    }

    private fun interpolate(a: LatLng, b: LatLng, t: Double): LatLng {
        return LatLng(a.latitude + (b.latitude - a.latitude) * t, a.longitude + (b.longitude - a.longitude) * t)
    }

    private suspend fun tick() {
        elapsed++
        val rs = RobotRepository.currentState()
        if (rs.controlDirection.isNotEmpty()) tickManual() else tickAuto()
        tickTelemetry()
        tickBattery()
    }

    private fun tickManual() {
        val rs = RobotRepository.currentState()
        val dir = rs.controlDirection
        val head = HEADING_ANGLES[dir] ?: 0
        val rad = head * PI / 180
        val speedMs = rs.controlSpeed
        val dist = speedMs
        val cosLat = cos(rs.latitude * PI / 180)

        val dLat = dist * cos(rad) / M_PER_DEG_LAT
        val dLng = dist * sin(rad) / (M_PER_DEG_LAT * cosLat)
        RobotRepository.updateLocation(rs.latitude + dLat, rs.longitude + dLng, head.toDouble(), speedMs * RobotRepository.MS_TO_KMH)
        RobotRepository.updateStatus(mileage = rs.mileage + dist / 1000, fix = 3, satellites = 12, locSrc = "gps")
        pulseCount += (speedMs * 100).toInt()
        RobotRepository.updateTelemetry(pulseCount, pulseCount + (0..5).random(), STEER_CENTER)
    }

    private fun tickAuto() {
        val rs = RobotRepository.currentState()
        val path = rs.path
        if (path.size < 2) return
        val nextIndex = if (forward) pathIndex + 1 else pathIndex - 1
        if (nextIndex < 0 || nextIndex >= path.size) { forward = !forward; pathIndex = if (forward) 0 else path.size - 1; return }
        val from = path[pathIndex]; val to = path[nextIndex]
        val segDist = distMeters(from, to)
        val speedMs = if (rs.controlDirection.isNotEmpty()) rs.controlSpeed else DEFAULT_AUTO_SPEED_MS
        val stepRatio = if (segDist > 0) speedMs / segDist else 0.0
        progress += stepRatio
        if (progress >= 1) { progress = 0.0; pathIndex = nextIndex }
        val pos = interpolate(from, to, minOf(progress, 1.0))
        val rawHeading = (atan2((to.longitude - from.longitude) * (if (forward) 1 else -1), (to.latitude - from.latitude) * (if (forward) 1 else -1)) * 180 / PI + 90)
        val heading = ((rawHeading % 360) + 360) % 360
        RobotRepository.updateLocation(pos.latitude, pos.longitude, heading, speedMs * RobotRepository.MS_TO_KMH)
        RobotRepository.updateStatus(mileage = rs.mileage + speedMs / 1000)
    }

    private fun tickTelemetry() {
        val rs = RobotRepository.currentState()
        pulseCount += (rs.controlSpeed * ENCODER_PULSE_RATE).toInt()
        val enc1 = pulseCount
        val isLeft = rs.controlDirection.contains("left"); val isRight = rs.controlDirection.contains("right")
        val enc2 = pulseCount + (if (isLeft) -20 else if (isRight) 20 else (0..5).random())
        val steer = if (isLeft) 120 else if (isRight) 180 else STEER_CENTER
        RobotRepository.updateTelemetry(enc1, enc2, steer)
        if (elapsed % 30 == 0) {
            val signals = listOf("强", "强", "强", "中", "弱")
            RobotRepository.updateStatus(signal = signals.random())
        }
    }

    private fun tickBattery() {
        val rs = RobotRepository.currentState()
        if (elapsed % 60 == 0 && rs.battery > BATTERY_MIN) RobotRepository.updateStatus(battery = rs.battery - 1)
    }

    fun start() {
        if (timer != null) return
        val rs = RobotRepository.currentState()
        RobotRepository.setConnected(true)
        RobotRepository.setMode("simulation")
        RobotRepository.updateStatus(status = "running")
        RobotRepository.updateControlSpeed(DEFAULT_AUTO_SPEED_MS)
        pathIndex = 0; progress = 0.0; forward = true; elapsed = 0
        timer = scope.launch {
            while (isActive) { tick(); delay(TICK_MS) }
        }
    }

    fun stop() {
        timer?.cancel(); timer = null
        RobotRepository.setConnected(false)
        RobotRepository.updateControlSpeed(0.0)
        RobotRepository.updateStatus(status = "online")
        RobotRepository.updateTelemetry(0, 0, 150)
    }

    fun setDirection(dir: String) {
        val rs = RobotRepository.currentState()
        if (dir == "stop" || dir.isEmpty()) {
            RobotRepository.updateControlDirection("", 0.0)
            RobotRepository.updateTelemetry(pulseCount, pulseCount, 150)
        } else {
            val speed = 0.8 + Math.random() * 0.6
            RobotRepository.updateControlDirection(dir, speed * 3.6)
        }
    }
}
